services/evm/executor.go

```diff
@@ -0,0 +1,186 @@
+
+	stateDB, err := state.NewStateDB(ledger, storageAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BlockExecutor{
+		emulator: emulator.NewEmulator(ledger, storageAddress),
+		StateDB:  stateDB,
+		chainID:  chainID,
+		block:    block,
+		blocks:   blocks,
+		receipts: receipts,
+		logger:   logger,
+	}, nil
+}
+
+func (s *BlockExecutor) Run(
+	tx models.Transaction,
+	tracer *tracers.Tracer,
+) (*gethTypes.Receipt, error) {
+	l := s.logger.With().Str("tx-hash", tx.Hash().String()).Logger()
+	l.Info().Msg("executing new transaction")
+
+	receipt, err := s.receipts.GetByTransactionID(tx.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, err := s.blockContext(receipt)
+	ctx.Tracer = tracer
+	if err != nil {
+		return nil, err
+	}
+
+	bv, err := s.emulator.NewBlockView(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var res *types.Result
+
+	switch t := tx.(type) {
+	case models.DirectCall:
+		res, err = bv.DirectCall(t.DirectCall)
+	case models.TransactionCall:
+		res, err = bv.RunTransaction(t.Transaction)
+	default:
+		return nil, fmt.Errorf("invalid transaction type")
+	}
+
+	if err != nil {
+		// todo is this ok, the service would restart and retry?
+		return nil, err
+	}
+
+	// we should never produce invalid transaction, since if the transaction was emitted from the evm core
+	// it must have either been successful or failed, invalid transactions are not emitted
+	if res.Invalid() {
+		return nil, fmt.Errorf("invalid transaction %s: %w", tx.Hash(), res.ValidationError)
+	}
+
+	// increment values as part of a virtual block
+	s.gasUsed += res.GasConsumed
+	s.txIndex++
+
+	l.Debug().Msg("transaction executed successfully")
+
+	return res.LightReceipt().ToReceipt(), nil
+}
+
+func (s *BlockExecutor) Call(
+	from common.Address,
+	data []byte,
+	tracer *tracers.Tracer,
+) (*types.Result, error) {
+	ctx, err := s.blockContext(nil)
+	ctx.Tracer = tracer
+	if err != nil {
+		return nil, err
+	}
+
+	bv, err := s.emulator.NewBlockView(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tx := &gethTypes.Transaction{}
+	if err := tx.UnmarshalBinary(data); err != nil {
+		return nil, err
+	}
+
+	return bv.DryRunTransaction(tx, from)
+}
+
+// blockContext produces a context that is used by the block view during the execution.
+// It can be used for transaction execution and calls. Receipt is not required when
+// producing the context for calls.
+func (s *BlockExecutor) blockContext(receipt *models.Receipt) (types.BlockContext, error) {
+	ctx := types.BlockContext{
+		ChainID:                types.EVMChainIDFromFlowChainID(s.chainID),
+		BlockNumber:            s.block.Height,
+		BlockTimestamp:         s.block.Timestamp,
+		DirectCallBaseGasUsage: types.DefaultDirectCallBaseGasUsage, // todo check
+		DirectCallGasPrice:     types.DefaultDirectCallGasPrice,
+		GasFeeCollector:        types.CoinbaseAddress,
+		GetHashFunc: func(n uint64) common.Hash {
+			b, err := s.blocks.GetByHeight(n)
+			if err != nil {
+				panic(err)
+			}
```

Comment:
We might need to change this to return empty if a higher block than the current block is passed. since the blocks might include the future blocks but for the sake of execution during a block, the future block hash should not be available. similarly for the block heights lower than the threshold (256?) we should return empty.

Reply:
Good catch :+1: Updated in https://github.com/onflow/flow-evm-gateway/pull/605/commits/0a7b3eca7e6dafbfcb24d0fbee45fb071515dd4f